front/components/assistant/conversation/AgentMessage.tsx

```diff
@@ -578,6 +516,21 @@ export function AgentMessage({
                     updateActiveReferences,
                     setHoveredReference: setLastHoveredReference,
                   }}
+                  customRenderer={{
+                    visualization: (code, complete, lineStart) => {
+                      return (
+                        <VisualizationActionIframe
+                          owner={owner}
+                          visualization={{
+                            code,
+                            complete,
+                            identifier: `viz-${agentMessage.sId}-${lineStart}`,
+                          }}
+                          key={`viz-${agentMessage.sId}-${lineStart}`}
+                        />
+                      );
+                    },
+                  }}
```

Comment:
You should memoize here. If you pass an anonymous function directly as a prop to a React component, it will create a new instance of that function every time the parent component re-renders (which is the case here because of the streaming). This will cause unnecessary re-rendering of child components because React sees the function as a new prop due to its reference changing.